# Проект "Книжный сервис"

<h3>Содержание проекта<span class="tocSkip"></span></h3>
<div class="toc"><ul class="toc-item"><li><ul class="toc-item"><li><span><a href="#Описание-проекта" data-toc-modified-id="Описание-проекта-0.1"><span class="toc-item-num">0.1&nbsp;&nbsp;</span>Описание проекта</a></span></li><li><span><a href="#Задачи-проекта" data-toc-modified-id="Задачи-проекта-0.2"><span class="toc-item-num">0.2&nbsp;&nbsp;</span>Задачи проекта</a></span></li><li><span><a href="#Описание-данных" data-toc-modified-id="Описание-данных-0.3"><span class="toc-item-num">0.3&nbsp;&nbsp;</span>Описание данных</a></span></li><li><span><a href="#Задания" data-toc-modified-id="Задания-0.4"><span class="toc-item-num">0.4&nbsp;&nbsp;</span>Задания</a></span></li></ul></li><li><span><a href="#Подключение-к-базе" data-toc-modified-id="Подключение-к-базе-1"><span class="toc-item-num">1&nbsp;&nbsp;</span>Подключение к базе</a></span><ul class="toc-item"><li><span><a href="#Импорт-библиотек" data-toc-modified-id="Импорт-библиотек-1.1"><span class="toc-item-num">1.1&nbsp;&nbsp;</span>Импорт библиотек</a></span></li><li><span><a href="#Создание-подключения" data-toc-modified-id="Создание-подключения-1.2"><span class="toc-item-num">1.2&nbsp;&nbsp;</span>Создание подключения</a></span></li></ul></li><li><span><a href="#Исследование-таблиц" data-toc-modified-id="Исследование-таблиц-2"><span class="toc-item-num">2&nbsp;&nbsp;</span>Исследование таблиц</a></span><ul class="toc-item"><li><span><a href="#Таблица-books" data-toc-modified-id="Таблица-books-2.1"><span class="toc-item-num">2.1&nbsp;&nbsp;</span>Таблица books</a></span></li><li><span><a href="#Таблица-authors" data-toc-modified-id="Таблица-authors-2.2"><span class="toc-item-num">2.2&nbsp;&nbsp;</span>Таблица authors</a></span></li><li><span><a href="#Таблица-publishers" data-toc-modified-id="Таблица-publishers-2.3"><span class="toc-item-num">2.3&nbsp;&nbsp;</span>Таблица publishers</a></span></li><li><span><a href="#Таблица-ratings" data-toc-modified-id="Таблица-ratings-2.4"><span class="toc-item-num">2.4&nbsp;&nbsp;</span>Таблица ratings</a></span></li><li><span><a href="#Таблица-reviews" data-toc-modified-id="Таблица-reviews-2.5"><span class="toc-item-num">2.5&nbsp;&nbsp;</span>Таблица reviews</a></span></li></ul></li><li><span><a href="#Задания" data-toc-modified-id="Задания-3"><span class="toc-item-num">3&nbsp;&nbsp;</span>Задания</a></span><ul class="toc-item"><li><span><a href="#Книги-после-1-января-2000-года" data-toc-modified-id="Книги-после-1-января-2000-года-3.1"><span class="toc-item-num">3.1&nbsp;&nbsp;</span>Книги после 1 января 2000 года</a></span><ul class="toc-item"><li><span><a href="#Решение" data-toc-modified-id="Решение-3.1.1"><span class="toc-item-num">3.1.1&nbsp;&nbsp;</span>Решение</a></span></li><li><span><a href="#Вывод" data-toc-modified-id="Вывод-3.1.2"><span class="toc-item-num">3.1.2&nbsp;&nbsp;</span>Вывод</a></span></li></ul></li><li><span><a href="#Количество-обзоров-и-средняя-оценка" data-toc-modified-id="Количество-обзоров-и-средняя-оценка-3.2"><span class="toc-item-num">3.2&nbsp;&nbsp;</span>Количество обзоров и средняя оценка</a></span><ul class="toc-item"><li><span><a href="#Подготовка-к-решению" data-toc-modified-id="Подготовка-к-решению-3.2.1"><span class="toc-item-num">3.2.1&nbsp;&nbsp;</span>Подготовка к решению</a></span></li><li><span><a href="#Решение" data-toc-modified-id="Решение-3.2.2"><span class="toc-item-num">3.2.2&nbsp;&nbsp;</span>Решение</a></span></li><li><span><a href="#Прверка" data-toc-modified-id="Прверка-3.2.3"><span class="toc-item-num">3.2.3&nbsp;&nbsp;</span>Прверка</a></span></li><li><span><a href="#Вывод" data-toc-modified-id="Вывод-3.2.4"><span class="toc-item-num">3.2.4&nbsp;&nbsp;</span>Вывод</a></span></li></ul></li><li><span><a href="#Издательство-с-наибольшим-числом-книг" data-toc-modified-id="Издательство-с-наибольшим-числом-книг-3.3"><span class="toc-item-num">3.3&nbsp;&nbsp;</span>Издательство с наибольшим числом книг</a></span><ul class="toc-item"><li><span><a href="#Подготовка-к-решению" data-toc-modified-id="Подготовка-к-решению-3.3.1"><span class="toc-item-num">3.3.1&nbsp;&nbsp;</span>Подготовка к решению</a></span></li><li><span><a href="#Решение" data-toc-modified-id="Решение-3.3.2"><span class="toc-item-num">3.3.2&nbsp;&nbsp;</span>Решение</a></span></li><li><span><a href="#Проверка-изданий-&quot;без-книг&quot;" data-toc-modified-id="Проверка-изданий-&quot;без-книг&quot;-3.3.3"><span class="toc-item-num">3.3.3&nbsp;&nbsp;</span>Проверка изданий "без книг"</a></span></li><li><span><a href="#Вывод" data-toc-modified-id="Вывод-3.3.4"><span class="toc-item-num">3.3.4&nbsp;&nbsp;</span>Вывод</a></span></li></ul></li><li><span><a href="#Автор-с-самой-высокой-средней-оценкой" data-toc-modified-id="Автор-с-самой-высокой-средней-оценкой-3.4"><span class="toc-item-num">3.4&nbsp;&nbsp;</span>Автор с самой высокой средней оценкой</a></span><ul class="toc-item"><li><span><a href="#Подготовка-к-решению" data-toc-modified-id="Подготовка-к-решению-3.4.1"><span class="toc-item-num">3.4.1&nbsp;&nbsp;</span>Подготовка к решению</a></span></li><li><span><a href="#Решение" data-toc-modified-id="Решение-3.4.2"><span class="toc-item-num">3.4.2&nbsp;&nbsp;</span>Решение</a></span></li><li><span><a href="#Выводы" data-toc-modified-id="Выводы-3.4.3"><span class="toc-item-num">3.4.3&nbsp;&nbsp;</span>Выводы</a></span></li></ul></li><li><span><a href="#Обзоры-пользователей" data-toc-modified-id="Обзоры-пользователей-3.5"><span class="toc-item-num">3.5&nbsp;&nbsp;</span>Обзоры пользователей</a></span><ul class="toc-item"><li><span><a href="#Подготовка-к-решению" data-toc-modified-id="Подготовка-к-решению-3.5.1"><span class="toc-item-num">3.5.1&nbsp;&nbsp;</span>Подготовка к решению</a></span></li><li><span><a href="#Решение" data-toc-modified-id="Решение-3.5.2"><span class="toc-item-num">3.5.2&nbsp;&nbsp;</span>Решение</a></span></li><li><span><a href="#Проверка" data-toc-modified-id="Проверка-3.5.3"><span class="toc-item-num">3.5.3&nbsp;&nbsp;</span>Проверка</a></span></li><li><span><a href="#Выводы" data-toc-modified-id="Выводы-3.5.4"><span class="toc-item-num">3.5.4&nbsp;&nbsp;</span>Выводы</a></span></li></ul></li></ul></li><li><span><a href="#Выводы-по-заданиям-и-предложения" data-toc-modified-id="Выводы-по-заданиям-и-предложения-4"><span class="toc-item-num">4&nbsp;&nbsp;</span>Выводы по заданиям и предложения</a></span><ul class="toc-item"><li><span><a href="#Выводы" data-toc-modified-id="Выводы-4.1"><span class="toc-item-num">4.1&nbsp;&nbsp;</span>Выводы</a></span></li><li><span><a href="#Предложения" data-toc-modified-id="Предложения-4.2"><span class="toc-item-num">4.2&nbsp;&nbsp;</span>Предложения</a></span></li></ul></li></ul></div>

### Описание проекта

Коронавирус застал мир врасплох, изменив привычный порядок вещей. В свободное время жители городов больше не выходят на улицу, не посещают кафе и торговые центры. Зато стало больше времени для книг. Это заметили стартаперы — и бросились создавать приложения для тех, кто любит читать.

Ваша компания решила быть на волне и купила крупный сервис для чтения книг по подписке. Ваша первая задача как аналитика — проанализировать базу данных.
В ней — информация о книгах, издательствах, авторах, а также пользовательские обзоры книг. Эти данные помогут сформулировать ценностное предложение для нового продукта.

### Задачи проекта

**Цель проекта:** изучить данные и сформулировать предложения по новому продукту. 

**Задачи проекта:** 
1. Проанализировать информацию о книгах, авторах и издательствах, а также пользовательские обзоры книг 
2. Сформулировать предложения по продукту 

### Описание данных

**Таблица `books`**

Содержит данные о книгах:

- `book_id` — идентификатор книги;
- `author_id` — идентификатор автора;
- `title` — название книги;
- `num_pages` — количество страниц;
- `publication_date` — дата публикации книги;
- `publisher_id` — идентификатор издателя.

**Таблица `authors`**

Содержит данные об авторах:

- `author_id` — идентификатор автора;
- `author` — имя автора.

**Таблица `publishers`**

Содержит данные об издательствах:

- `publisher_id` — идентификатор издательства;
- `publisher` — название издательства;

**Таблица `ratings`**

Содержит данные о пользовательских оценках книг:

- `rating_id` — идентификатор оценки;
- `book_id` — идентификатор книги;
- `username` — имя пользователя, оставившего оценку;
- `rating` — оценка книги.

**Таблица `reviews`**

Содержит данные о пользовательских обзорах на книги:

- `review_id` — идентификатор обзора;
- `book_id` — идентификатор книги;
- `username` — имя пользователя, написавшего обзор;
- `text` — текст обзора.

### Задания

1. Посчитайте, сколько книг вышло после 1 января 2000 года;
2. Для каждой книги посчитайте количество обзоров и среднюю оценку;
3. Определите издательство, которое выпустило наибольшее число книг толще 50 страниц — так вы исключите из анализа брошюры;
4. Определите автора с самой высокой средней оценкой книг — учитывайте только книги с 50 и более оценками;
5. Посчитайте среднее количество обзоров от пользователей, которые поставили больше 50 оценок.

## Подключение к базе 

### Импорт библиотек

In [1]:
import pandas as pd
from sqlalchemy import create_engine

### Создание подключения

In [2]:
# устанавливаем параметры
db_config = {'user': 'praktikum_student', # имя пользователя
 'pwd': 'Sdf4$2;d-d30pp', # пароль
 'host': 'rc1b-wcoijxj3yxfsf3fs.mdb.yandexcloud.net',
 'port': 6432, # порт подключения
 'db': 'data-analyst-final-project-db'} # название базы данных
connection_string = 'postgresql://{}:{}@{}:{}/{}'.format(db_config['user'],
 db_config['pwd'],
 db_config['host'],
 db_config['port'],
 db_config['db'])

# сохраняем коннектор
engine = create_engine(connection_string, connect_args={'sslmode':'require'}) 

## Исследование таблиц

**Определим функцию для запросов из БД:**

In [3]:
def select(query): 
    return pd.io.sql.read_sql(query, con=engine) 

### Таблица books

**Таблица `books`**

Содержит данные о книгах:

- `book_id` — идентификатор книги;
- `author_id` — идентификатор автора;
- `title` — название книги;
- `num_pages` — количество страниц;
- `publication_date` — дата публикации книги;
- `publisher_id` — идентификатор издателя.

In [4]:
books = '''
SELECT *
FROM books
'''
select(books + 'LIMIT 5')

book_id  author_id                                              title  \
0        1        546                                       'Salem's Lot   
1        2        465                 1 000 Places to See Before You Die   
2        3        407  13 Little Blue Envelopes (Little Blue Envelope...   
3        4         82  1491: New Revelations of the Americas Before C...   
4        5        125                                               1776   

   num_pages publication_date  publisher_id  
0        594       2005-11-01            93  
1        992       2003-05-22           336  
2        322       2010-12-21           135  
3        541       2006-10-10           309  
4        386       2006-07-04           268

In [5]:
sql = '''
SELECT count(t.book_id) AS books_count,
       count(DISTINCT author_id) AS authors_count,
       count(DISTINCT publisher_id) AS publishers_count
FROM books t
'''
select(sql)

books_count  authors_count  publishers_count
0         1000            636               340

**Размер таблицы:** 6 столбцов и 1000 строк 

Отметим, что уникальных авторов в таблице 636, а уникальных изданий - 340

### Таблица authors

**Таблица `authors`**

Содержит данные об авторах:

- `author_id` — идентификатор автора;
- `author` — имя автора.

In [6]:
authors = '''
SELECT *
FROM authors
'''
select(authors + 'LIMIT 5')

author_id                          author
0          1                      A.S. Byatt
1          2  Aesop/Laura Harris/Laura Gibbs
2          3                 Agatha Christie
3          4                   Alan Brennert
4          5        Alan Moore/David   Lloyd

In [7]:
sql = '''
SELECT COUNT(t.author_id) AS authors_count
FROM authors t
'''
select(sql)

authors_count
0            636

**Размер таблицы:** 2 столбца и 636 строк

Так как авторов меньше, чем книг, можно сделать закономерный вывод о том, что в таблице есть авторы, которые написали больше одной книги. При этом число авторов таблиц `authors` и `books` совпадает   

### Таблица publishers

**Таблица `publishers`**

Содержит данные об издательствах:

- `publisher_id` — идентификатор издательства;
- `publisher` — название издательства;

In [8]:
publishers = '''
SELECT *
FROM publishers
'''
select(publishers + 'LIMIT 5')

publisher_id                          publisher
0             1                                Ace
1             2                           Ace Book
2             3                          Ace Books
3             4                      Ace Hardcover
4             5  Addison Wesley Publishing Company

In [9]:
sql = '''
SELECT count(t.publisher_id) AS publishers_count
FROM publishers t
'''
select(sql)

publishers_count
0               340

**Размер таблицы:** 2 столбца и 340 строк

Число издательств таблиц `publishers` и `books` совпадает   

### Таблица ratings

**Таблица `ratings`**

Содержит данные о пользовательских оценках книг:

- `rating_id` — идентификатор оценки;
- `book_id` — идентификатор книги;
- `username` — имя пользователя, оставившего оценку;
- `rating` — оценка книги.

In [10]:
ratings = '''
SELECT *
FROM ratings
'''
select(ratings + 'LIMIT 5')

rating_id  book_id       username  rating
0          1        1     ryanfranco       4
1          2        1  grantpatricia       2
2          3        1   brandtandrea       5
3          4        2       lorichen       3
4          5        2    mariokeller       2

In [11]:
sql = '''
SELECT count(t.rating_id) AS ratings_count,
       count(DISTINCT t.book_id) AS books_count
FROM ratings t
'''
select(sql)

ratings_count  books_count
0           6456         1000

**Размер таблицы:** 4 столбца и 6456 строк

Оценка есть у каждой книги

### Таблица reviews 

**Таблица `reviews`**

Содержит данные о пользовательских обзорах на книги:

- `review_id` — идентификатор обзора;
- `book_id` — идентификатор книги;
- `username` — имя пользователя, написавшего обзор;
- `text` — текст обзора.

In [12]:
reviews = '''
SELECT *
FROM reviews
'''
select(reviews + 'LIMIT 5')

review_id  book_id       username  \
0          1        1   brandtandrea   
1          2        1     ryanfranco   
2          3        2       lorichen   
3          4        3  johnsonamanda   
4          5        3    scotttamara   

                                                text  
0  Mention society tell send professor analysis. ...  
1  Foot glass pretty audience hit themselves. Amo...  
2  Listen treat keep worry. Miss husband tax but ...  
3  Finally month interesting blue could nature cu...  
4  Nation purpose heavy give wait song will. List...

In [13]:
sql = '''
SELECT count(t.review_id) AS reviews_count,
       count(DISTINCT t.book_id) AS books_count
FROM reviews t
'''
select(sql)

reviews_count  books_count
0           2793          994

**Размер таблицы:** 4 столбца и 2793 строки

При этом стоит заметить, что на 6 книг нет пользовательских обзоров. Это может быть важно при дальнейшем анализе 

<div class="alert alert-success">

**Комментарий ревьюера**
     
Здорово, что данные были изучены перед анализом 👍🏻

##  Задания

1. Посчитайте, сколько книг вышло после 1 января 2000 года;
2. Для каждой книги посчитайте количество обзоров и среднюю оценку;
3. Определите издательство, которое выпустило наибольшее число книг толще 50 страниц — так вы исключите из анализа брошюры;
4. Определите автора с самой высокой средней оценкой книг — учитывайте только книги с 50 и более оценками;
5. Посчитайте среднее количество обзоров от пользователей, которые поставили больше 50 оценок.

### Книги после 1 января 2000 года

Посчитайте, сколько книг вышло после 1 января 2000 года

#### Решение

In [14]:
task_1 = '''
SELECT COUNT(t.book_id) AS books_count
FROM books t
WHERE publication_date > '2000-01-01'
'''
select(task_1)

books_count
0          819

#### Вывод

После 1 января 2000 года вышло 819 книг из 1000, представленных в таблице

###  Количество обзоров и средняя оценка

Для каждой книги посчитайте количество обзоров и среднюю оценку

**Какие данные нам понадобятся:**
1. `books.book_id`
2. `books.title`
3. `reviews.review_id`
4. `ratings.rating`

**Структура будущей таблицы:**
1. `books.book_id`
2. `books.title`
3. `reviews.reviews_count`
4. `ratings.rating_avg`

**Последовательность объединений таблиц:**
1. `reviews` + `ratings` -> `ratings_reviews`
2. `books` + `ratings_reviews` -> `task_final`

#### Подготовка к решению

**1. Найдем число обзоров на каждыую книгу**

In [15]:
sql = '''
SELECT t1.book_id,
       count(t1.review_id) AS reviews_count
FROM reviews t1
GROUP BY t1.book_id
'''
select(sql + 'LIMIT 5')

book_id  reviews_count
0      652              2
1      273              2
2       51              5
3      951              2
4      839              4

**2. Найдем среднюю оценку каждой книги**

In [16]:
sql = '''
SELECT t2.book_id,
       round(avg(t2.rating), 2) AS rating_avg
FROM ratings t2
GROUP BY t2.book_id
'''
select(sql + 'LIMIT 5')

book_id  rating_avg
0      652        4.50
1      273        4.50
2       51        4.25
3      951        4.00
4      839        4.29

**3. Найдем средниюю оценку и количество обзоров на каждую книгу**

In [17]:
sql = '''
WITH new_reviews AS
  (SELECT t1.book_id,
          count(t1.review_id) AS reviews_count
   FROM reviews t1
   GROUP BY t1.book_id),
     new_ratings AS
  (SELECT t2.book_id,
          round(avg(t2.rating), 2) AS rating_avg
   FROM ratings t2
   GROUP BY t2.book_id),
     ratings_reviews AS
  (SELECT t3.book_id,
          t3.rating_avg,
          t4.reviews_count
   FROM (new_ratings t3
         LEFT JOIN new_reviews t4 ON t3.book_id = t4.book_id))
SELECT *
FROM ratings_reviews
'''
select(sql + 'LIMIT 5')

book_id  rating_avg  reviews_count
0        1        3.67              2
1        2        2.50              1
2        3        4.67              3
3        4        4.50              2
4        5        4.00              4

**4. Проверим, что в итоговой таблице содержатся все книги**

In [18]:
sql = '''
WITH new_reviews AS
  (SELECT t1.book_id,
          count(t1.review_id) AS reviews_count
   FROM reviews t1
   GROUP BY t1.book_id),
     new_ratings AS
  (SELECT t2.book_id,
          round(avg(t2.rating), 2) AS rating_avg
   FROM ratings t2
   GROUP BY t2.book_id),
     ratings_reviews AS
  (SELECT t3.book_id,
          t3.rating_avg,
          t4.reviews_count
   FROM (new_ratings t3
         LEFT JOIN new_reviews t4 ON t3.book_id = t4.book_id))
SELECT count(t.book_id)
FROM ratings_reviews t
'''
select(sql)

count
0   1000

#### Решение

In [19]:
task_2 = '''
WITH new_reviews AS
  (SELECT t1.book_id,
          count(t1.review_id) AS reviews_count
   FROM reviews t1
   GROUP BY t1.book_id),
     new_ratings AS
  (SELECT t2.book_id,
          round(avg(t2.rating), 2) AS rating_avg
   FROM ratings t2
   GROUP BY t2.book_id),
     ratings_reviews AS
  (SELECT t3.book_id,
          t3.rating_avg,
          t4.reviews_count
   FROM (new_ratings t3
         LEFT JOIN new_reviews t4 ON t3.book_id = t4.book_id))
SELECT t5.book_id,
       t5.title,
       t6.reviews_count,
       t6.rating_avg
FROM (books t5
      LEFT JOIN ratings_reviews t6 ON t5.book_id = t6.book_id)
'''
select(task_2 + 'where t6.reviews_count is not null order by t6.reviews_count desc limit 10')

book_id                                              title  reviews_count  \
0      948                            Twilight (Twilight  #1)              7   
1      696             The Da Vinci Code (Robert Langdon  #2)              6   
2      207                                    Eat  Pray  Love              6   
3      627                                      The Alchemist              6   
4      673                             The Catcher in the Rye              6   
5      695  The Curious Incident of the Dog in the Night-Time              6   
6      302  Harry Potter and the Prisoner of Azkaban (Harr...              6   
7      299  Harry Potter and the Chamber of Secrets (Harry...              6   
8      497                          Outlander (Outlander  #1)              6   
9      656                                     The Book Thief              6   

   rating_avg  
0        3.66  
1        3.83  
2        3.40  
3        3.79  
4        3.83  
5        4.08  
6        4.41  
7        4.29  
8        4.13  
9        4.26

1. Максимальное число обзоров получила книга с id 948 и названием `Twilight (Twilight #1)` - целых 7. 6 обзоров получило уже значительно бОльшее число книг
2. Однако книги с большим числом оценок далеко не всегда имеют высокие оценки. Так, первые 5 позиций получили среднюю оценку меншье 4 

#### Прверка

In [20]:
sql = '''
WITH new_reviews AS
  (SELECT t1.book_id,
          count(t1.review_id) AS reviews_count
   FROM reviews t1
   GROUP BY t1.book_id),
     new_ratings AS
  (SELECT t2.book_id,
          round(avg(t2.rating), 2) AS rating_avg
   FROM ratings t2
   GROUP BY t2.book_id),
     ratings_reviews AS
  (SELECT t3.book_id,
          t3.rating_avg,
          t4.reviews_count
   FROM (new_ratings t3
         LEFT JOIN new_reviews t4 ON t3.book_id = t4.book_id)),
     final_tab AS
  (SELECT t5.book_id,
          t5.title,
          t6.reviews_count,
          t6.rating_avg
   FROM (books t5
         LEFT JOIN ratings_reviews t6 ON t5.book_id = t6.book_id))
SELECT count(t.book_id)
FROM final_tab t
'''
select(sql)

count
0   1000

Все книги попали в итоговую таблицу

In [21]:
select(sql + 'where t.reviews_count is null')

count
0      6

При этом как и предполагалось, 6 книг остались без обзоров

#### Вывод

1. Максимальное число обзоров получила книга с id 948 и названием `Twilight (Twilight #1)` - целых 7. 6 обзоров получило уже значительно бОльшее число книг
2. Однако книги с большим числом оценок далеко не всегда имеют высокие оценки. Так, первые 5 позиций получили среднюю оценку меншье 4 

### Издательство с наибольшим числом книг

Определите издательство, которое выпустило наибольшее число книг толще 50 страниц — так вы исключите из анализа брошюры

**Какие данные нам понадобятся:**
1. `publishers.publisher_id`
2. `publishers.publisher`
3. `books.num_pages`
4. `books.publisher_id`
5. `books.book_id`

**Структура будущей таблицы:**
1. `publishers.publisher_id`
2. `publishers.publisher`
3. `books.books_count`

**Последовательность объединений таблиц:**
1. `books` - `publishers` -> `books_by_publisher`

#### Подготовка к решению

**1. Найдем число книг, опубликованных в каждом издании, у которых больше 50 страниц**

In [22]:
sql = '''
SELECT t1.publisher_id,
       count(t1.book_id) AS books_count
FROM books t1
WHERE t1.num_pages > 50
GROUP BY t1.publisher_id
'''
select(sql + 'LIMIT 5')

publisher_id  books_count
0            87            1
1           184            1
2           116           25
3           273            1
4            51            2

#### Решение

In [25]:
task_3 = '''
WITH books_by_publisher AS
  (SELECT t1.publisher_id,
          count(t1.book_id) AS books_count
   FROM books t1
   WHERE t1.num_pages > 50
   GROUP BY t1.publisher_id)
SELECT t2.publisher_id,
       t2.publisher,
       t3.books_count
FROM (publishers t2
      LEFT JOIN books_by_publisher t3 ON t2.publisher_id = t3.publisher_id)
'''

output_cond = '''
WHERE t3.books_count IS NOT NULL
ORDER BY t3.books_count DESC
LIMIT 5
'''
select(task_3 + output_cond)

publisher_id                 publisher  books_count
0           212             Penguin Books           42
1           309                   Vintage           31
2           116  Grand Central Publishing           25
3           217          Penguin Classics           24
4            35                    Bantam           19

1. Наибольшее число опубликованных книг наблюдается у издательства `Penguin Books	`
2. Отметим, что некоторые издания остались "без книг", т.к. все книги в них 

#### Проверка изданий "без книг"

In [26]:
select(task_3 + 'where t3.books_count is null')

publisher_id                         publisher books_count
0           270  Simon & Schuster Childrens Books        None
1            25  Atheneum Books for Young Readers        None
2           288                        Tanglewood        None
3           112                      Golden Books        None
4           113                     Golden/Disney        None
5            55                  Brilliance Audio        None

In [27]:
sql='''
SELECT max(t.num_pages)
FROM books t
WHERE t.publisher_id = 270
'''
select(sql)

max
0   14

Как видно из тестовой проверки, самая "толстая" книга у издания с id 270 - всего 14 страниц. Значит отбор данных произошел корректно

#### Вывод

1. Наибольшее число опубликованных книг наблюдается у издательства `Penguin Books	`
2. Отметим, что некоторые издания остались "без книг", т.к. все книги в них 

### Автор с самой высокой средней оценкой

Определите автора с самой высокой средней оценкой книг — учитывайте только книги с 50 и более оценками

**Какие данные нам понадобятся:**
1. `authors.author_id`
2. `authors.author`
3. `ratings.rating`
4. `ratings.book_id`
5. `ratings.rating_id`
5. `books.book_id`
6. `books.author_id`

**Структура будущей таблицы:**
1. `authors.author_id`
2. `authors.author`
3. `books_rating.avg_author_rating`

**Последовательность объединений таблиц:**
1. `books` (t1) - `ratings` (t2) -> `books_rating` (t3)
2. `authors` - `books_rating`

#### Подготовка к решению

**1. Найдем число оценок на каждую книгу**

In [28]:
sql = '''
SELECT t.book_id,
       count(t.rating_id) AS rating_count
FROM ratings t
GROUP BY t.book_id
HAVING count(t.rating_id) >= 50
'''
select(sql + 'limit 5')

book_id  rating_count
0       75            84
1      750            88
2      545            66
3      948           160
4      488            61

**2. Посмотрим, сколько всего книг имеют более 50 оценок**

In [29]:
sql = '''
WITH books_50 AS
  (SELECT t.book_id,
          count(t.rating_id) AS rating_count
   FROM ratings t
   GROUP BY t.book_id
   HAVING count(t.rating_id) >= 50)
SELECT count(t.book_id)
FROM books_50 t
'''
select(sql)

count
0     19

**3. Выведем таблицу со всеми оценками попавших в выборку книг**

In [30]:
sql = '''
WITH books_50 AS
  (SELECT t.book_id,
          count(t.rating_id) AS rating_count
   FROM ratings t
   GROUP BY t.book_id
   HAVING count(t.rating_id) >= 50)
SELECT t2.book_id,
       t2.rating
FROM (books_50 t1
      LEFT JOIN ratings t2 ON t1.book_id = t2.book_id)
'''
select(sql + 'LIMIT 10')

book_id  rating
0       75       5
1       75       4
2       75       4
3       75       4
4       75       4
5       75       4
6       75       5
7       75       2
8       75       3
9       75       3

**4. Найдем среднюю оценку всех книг, сгруппированных по автору**

In [31]:
sql = '''
WITH books_50 AS
  (SELECT t.book_id,
          count(t.rating_id) AS rating_count
   FROM ratings t
   GROUP BY t.book_id
   HAVING count(t.rating_id) >= 50),
     books_50_rating AS
  (SELECT t2.book_id,
          t2.rating
   FROM (books_50 t1
         INNER JOIN ratings t2 ON t1.book_id = t2.book_id))
SELECT t3.author_id,
       count(t4.rating) AS author_rating_count,
       avg(t4.rating) AS author_avg_rating
FROM (books t3
      INNER JOIN books_50_rating t4 ON t3.book_id = t4.book_id)
GROUP BY t3.author_id
'''
select(sql)

author_id  author_rating_count  author_avg_rating
0         554                  160           3.662500
1         630                   66           3.787879
2         498                   62           4.080645
3         469                   57           3.789474
4         236                  310           4.287097
5         311                   61           3.622951
6         240                  162           4.246914
7         402                   53           4.264151
8         195                   74           3.729730
9         372                   56           3.750000
10        621                   71           3.901408
11        106                  143           3.741259
12        235                   86           3.825581
13        376                   52           4.192308

#### Решение

In [32]:
task_4 = '''
WITH books_50 AS
  (SELECT t.book_id,
          count(t.rating_id) AS rating_count
   FROM ratings t
   GROUP BY t.book_id
   HAVING count(t.rating_id) >= 50),
     books_50_rating AS
  (SELECT t2.book_id,
          t2.rating
   FROM (books_50 t1
         INNER JOIN ratings t2 ON t1.book_id = t2.book_id)),
     top_authors AS
  (SELECT t3.author_id,
          avg(t4.rating) AS author_avg_rating
   FROM (books t3
         INNER JOIN books_50_rating t4 ON t3.book_id = t4.book_id)
   GROUP BY t3.author_id)
SELECT t5.author_id,
       t5.author,
       t6.author_avg_rating
FROM (authors t5
      INNER JOIN top_authors t6 ON t5.author_id = t6.author_id)
ORDER BY t6.author_avg_rating DESC
'''
select(task_4 + 'limit 5')

author_id                             author  author_avg_rating
0        236         J.K. Rowling/Mary GrandPré           4.287097
1        402  Markus Zusak/Cao Xuân Việt Khương           4.264151
2        240                     J.R.R. Tolkien           4.246914
3        376                  Louisa May Alcott           4.192308
4        498                       Rick Riordan           4.080645

#### Выводы

1. Всего 19 книг имеют более 50 оценок
2. Наибольшую среднюю оценку по всем книгам получили книги J.K. Rowling и Mary GrandPré

### Обзоры пользователей 

Посчитайте среднее количество обзоров от пользователей, которые поставили больше 50 оценок.

**Какие данные нам понадобятся:**
1. `ratings.rating_id`
2. `ratings.username`
3. `reviews.username`
4. `reviews.review_id`

**Структура будущей таблицы:**
1. `ratings.username`
2. `ratings.rating_count`
3. `reviews.reviews_count`

**Последовательность объединений таблиц:**
1. `ratings` (t1) - `reviews` (t2) -> `rating_review` (t3)

#### Подготовка к решению

**1. Найдем пользователей с числом оценок больше 50**

In [33]:
sql='''
SELECT t.username,
       count(t.rating_id) AS rating_count
FROM ratings t
GROUP BY t.username
HAVING count(t.rating_id) > 50
'''
select(sql)

username  rating_count
0     sfitzgerald            55
1  jennifermiller            53
2          xdavis            51
3          paul88            56
4      martinadam            56
5       richard89            55

Как видно, таких пользователей всего 6 

**2. Найдем число обзоров каждого пользователя**

In [34]:
sql = '''
SELECT t.username,
       count(t.review_id) AS review_count
FROM reviews t
GROUP BY t.username 
'''
select(sql + 'limit 5')

username  review_count
0      zjohnston            14
1        jnelson            21
2   lewisdesiree            19
3  patrickhudson            15
4          ulowe            16

#### Решение

In [35]:
task_5='''
WITH new_rating AS
  (SELECT t.username,
          count(t.rating_id) AS rating_count
   FROM ratings t
   GROUP BY t.username
   HAVING COUNT(t.rating_id) > 50),
     FINAL AS
  (SELECT t2.username,
          t1.rating_count,
          count(t2.review_id) AS review_count
   FROM (new_rating t1
         INNER JOIN reviews t2 ON t1.username = t2.username)
   GROUP BY t2.username,
            t1.rating_count)
SELECT ROUND(AVG(t.review_count), 2) AS avg_review_count
FROM FINAL t
'''
select(task_5)

avg_review_count
0             24.33

В среднем пользователи, которые поставили больше 50 оценок, написали по 24,3 обзора 

#### Проверка 

In [36]:
sql ='''
WITH new_rating AS
  (SELECT t.username,
          count(t.rating_id) AS rating_count
   FROM ratings t
   GROUP BY t.username
   HAVING COUNT(t.rating_id) > 50),
     final_tab AS
  (SELECT t2.username,
          t1.rating_count,
          count(t2.review_id) AS review_count
   FROM (new_rating t1
         INNER JOIN reviews t2 ON t1.username = t2.username)
   GROUP BY t2.username,
            t1.rating_count)
SELECT count(t.username)
FROM final_tab t
'''
select(sql)

count
0      6

В итогой таблице, как и ожидалось, осталось всего 6 пользователей

#### Выводы

1. В среднем пользователи, которые поставили больше 50 оценок, написали по 24,3 обзора
2. Однако таких пользователей в таблице всего 6 

## Выводы по заданиям и предложения

### Выводы

**Задание 1**

1. После 1 января 2000 года вышло 819 книг из 1000, представленных в таблице

**Задание 2**

1. Максимальное число обзоров получила книга с id 948 и названием `Twilight (Twilight #1)` - целых 7. 6 обзоров получило уже значительно бОльшее число книг
2. Однако книги с большим числом оценок далеко не всегда имеют высокие оценки. Так, первые 5 позиций получили среднюю оценку меншье 4 

**Задание 3**

1. Наибольшее число опубликованных книг наблюдается у издательства `Penguin Books	`
2. Отметим, что некоторые издания остались "без книг", т.к. все книги в них 

**Задание 4**

1. Всего 19 книг имеют более 50 оценок
2. Наибольшую среднюю оценку по всем книгам получили книги J.K. Rowling и Mary GrandPré

**Задание 5**

1. В среднем пользователи, которые поставили больше 50 оценок, написали по 24,3 обзора
2. Однако таких пользователей в таблице всего 6 


### Предложения

В целом, предложения могут быть следующие: 
1. Большое количество оценок и обзоров дают более объективную оценку. Однако эта оценка далеко не всегда выоская. В рекомендательную систему стоит включать книги с достаточно большим числом обзоров и оценок, но еще и со средней оценкой чуть выше среднего 
2. Рекомендации можно выстраивать на основе совпадения издательств. Возможно, книги издательств будут иметь схожие черты и нравиться пользователю 
3. Есть авторы с книгами-бестселлерами. У них очень высокие средние оценки. Возможно, стоит акцентировать внимание на этих авторах и их книгах 
4. В базе есть супер активные пользователи, которые ставят огромное число оценок и вместе с этим пишут огромное число обзоров. Возможно на основе общности обозреваемых ими книг также можно выстраивать некоторые рекомендации 